### ISSUE TITLE
[torch.compile] A potential OOB access in CUDA for attention model

### ISSUE THREAD

**Issue by YangChenyuan** [pytorch/pytorch#122381]
### 🐛 Describe the bug

For the following model, the result optimized by `torch.compile` on CUDA is totally wrong, which may out-of-bound access some data.

```py
import torch

torch.manual_seed(420)

class Model(torch.nn.Module):

    def __init__(self, d_model, dropout_p=0.1, inv_scale_factor=1e-08):
        super().__init__()
        self.d_model = d_model
        self.dropout_p = dropout_p
        self.inv_scale_factor = inv_scale_factor
        self.dropout = torch.nn.Dropout(dropout_p)

    def forward(self, query, key, value):
        qk = torch.matmul(query, key.transpose(-2, -1))
        scaled_qk = qk.div(self.inv_scale_factor)
        softmax_qk = scaled_qk.softmax(dim=-1)
        dropout_qk = self.dropout(softmax_qk)
        return dropout_qk

device = "cuda"
d_model = 1
func = Model(10).to(device)

query = torch.randn(1, 5, 10).to(device)

key = torch.randn(1, 5, 10).to(device)

value = torch.randn(1, 5, 10).to(device)

with torch.no_grad():
    naive_result = func(query.clone(), key.clone(), value.clone())

    func1 = torch.compile(func)
    jit_result = func1(query.clone(), key.clone(), value.clone())

    print(naive_result)
    print(jit_result)
    print(torch._dynamo.utils.counters["inductor"])
"""
tensor([[[0.0000, 0.0000, 0.0000, 1.1111, 0.0000],
         [0.0000, 0.0000, 1.1111, 0.0000, 0.0000],
         [0.0000, 0.0000, 0.0000, 0.0000, 1.1111],
         [0.0000, 0.0000, 0.0000, 1.1111, 0.0000],
         [0.0000, 0.0000, 1.1111, 0.0000, 0.0000]]], device='cuda:0')
tensor([[[0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00],
         [0.0000e+00, 0.0000e+00, 1.6732e+03, 0.0000e+00, 0.0000e+00],
         [0.0000e+00, 0.0000e+00, 0.0000e+00, 0.0000e+00, 4.0731e-10],
         [0.0000e+00, 0.0000e+00, 0.0000e+00, 3.4720e-03, 0.0000e+00],
         [0.0000e+00, 0.0000e+00, 4.0205e+06, 0.0000e+00, 0.0000e+00]]],
       device='cuda:0')
Counter({'pattern_matcher_count': 4, 'pattern_matcher_nodes': 4})
"""
```

This model has been minimized; therefore, if any operator is removed, it won't trigger the bug. Additionally, I've printed the patterns optimized by the Inductor to assist with debugging.

```
view (sum_1, [4]) Match(..., [], {'arg': sum_1, 'size': [4]}) CallFunction(aten.view.default, KeywordArg('arg'), KeywordArg('size'))
view_2 (None, [1, 5, 5]) Match(..., [], {'arg': bmm, 'size': [1, 5, 5]}) CallFunction(aten.view.default, KeywordArg('arg'), KeywordArg('size'))
view_1 (None, [1, 10, 5]) Match(..., [], {'arg': expand_1, 'size': [1, 10, 5]}) CallFunction(aten.view.default, KeywordArg('arg'), KeywordArg('size'))
view (None, [1, 5, 10]) Match(..., [], {'arg': expand, 'size': [1, 5, 10]}) CallFunction(aten.view.default, KeywordArg('arg'), KeywordArg('size'))
rand ([1, 5, 5],) Match(..., [[1, 5, 5]], {'dtype': torch.float32, 'device': device(type='cuda', index=0), 'pin_memory': False}) CallFunctionVarArgs(aten.rand.default)
```

It seems that this is **not** related to the `fuse_attention` optimization

### Versions

```
Collecting environment information...
PyTorch version: 2.3.0.dev20240301+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.1 LTS (x86_64)
GCC version: (Ubuntu 12.3.0-1ubuntu1~22.04) 12.3.0
Clang version: 11.0.0 (https://github.com/aflgo/aflgo.git fa125da5d70621daf7141c6279877c97708c8c1f)
CMake version: version 3.22.1
Libc version: glibc-2.35

Python version: 3.9.0 (default, Nov 15 2020, 14:28:56)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-6.5.0-26-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3060
Nvidia driver version: 525.147.05
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             24
On-line CPU(s) list:                0-23
Vendor ID:                          GenuineIntel
Model name:                         12th Gen Intel(R) Core(TM) i9-12900K
CPU family:                         6
Model:                              151
Thread(s) per core:                 2
Core(s) per socket:                 16
Socket(s):                          1
Stepping:                           2
CPU max MHz:                        5200.0000
CPU min MHz:                        800.0000
BogoMIPS:                           6374.40
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault cat_l2 cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect avx_vnni dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq tme rdpid movdiri movdir64b fsrm md_clear serialize pconfig arch_lbr ibt flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          640 KiB (16 instances)
L1i cache:                          768 KiB (16 instances)
L2 cache:                           14 MiB (10 instances)
L3 cache:                           30 MiB (1 instance)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-23
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pytorch-triton==3.0.0+901819d2b6
[pip3] torch==2.3.0.dev20240301+cu121
[pip3] torchaudio==2.2.0.dev20240301+cu118
[pip3] torchvision==0.18.0.dev20240301+cu118
[pip3] triton==2.2.0
[conda] numpy                     1.26.4                   pypi_0    pypi
[conda] pytorch-triton            3.0.0+901819d2b6          pypi_0    pypi
[conda] torch                     2.3.0.dev20240301+cu121          pypi_0    pypi
[conda] torchaudio                2.2.0.dev20240301+cu118          pypi_0    pypi
[conda] torchvision               0.18.0.dev20240301+cu118          pypi_0    pypi
[conda] triton                    2.2.0                    pypi_0    pypi

```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @bdhirsh @anijain2305 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire

**Comment by shunting314** [comment]
So there are 2 issues here:
1. by default inductor will use tl.rand (triton API) to generate random numbers. The generated random numbers are different to those generated by eager mode. If you want to test accuracy, you can force inductor to fallback to eager RNG by setting torch._inductor.config.fallback_random to True.  Also, you may want to reset the seed before both eager and compiled calls to make sure they use the same random numbers (for the dropout op).
2. besides that there is another issue related to FMA that cause accuracy problem. cc @Chillee 

If I disable both triton RNG and FMA, you test can pass.

For item2, Chillee mentioned that @peterbell10 is working on a PR.

**Comment by shunting314** [comment]
This issue has the same root cause as https://github.com/pytorch/pytorch/issues/122260 . Close it for duplication. A WIP PR is linked to the other issue.
